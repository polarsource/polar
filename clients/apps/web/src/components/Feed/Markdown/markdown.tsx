import { AdvertisementDisplay, Article } from '@polar-sh/sdk'
import React, { Fragment } from 'react'
import * as ReactIs from 'react-is'

export type RenderArticle = Pick<
  Article,
  | 'title'
  | 'body'
  | 'published_at'
  | 'byline'
  | 'organization'
  | 'slug'
  | 'is_preview'
>

export type BenefitAds = {
  benefitId: string
  ads: Array<AdvertisementDisplay>
}

// strictCreateElement removes unsupported types and attributes
export const wrapStrictCreateElement = (args: {
  article: RenderArticle
  showPaywalledContent?: boolean
  isSubscriber?: boolean
  paidArticlesBenefitName?: string
  defaultOverride?: React.FunctionComponent
  extraAllowedCustomComponents?: string[]
  adsContext?: BenefitAds[]
}): ((
  type: string | React.FunctionComponent<{}> | React.ComponentClass<{}, any>,
  props: JSX.IntrinsicAttributes | any,
  ...children: React.ReactNode[]
) => JSX.Element) => {
  let globalIdx = 0

  const strictCreateElement = (
    type: string | React.FunctionComponent<{}> | React.ComponentClass<{}, any>,
    props: JSX.IntrinsicAttributes | any,
    ...children: React.ReactNode[]
  ): JSX.Element => {
    const retNode = (node: React.ReactNode): JSX.Element => {
      if (!node) {
        return <></>
      }

      if (Array.isArray(node) && node.length === 0) {
        return <></>
      }

      if (Array.isArray(node) && node.length === 1) {
        return node[0]
      }

      if (Array.isArray(node)) {
        return (
          <Fragment>
            {node.map((ch) => (
              <Fragment key={globalIdx++}>{retNode(ch)}</Fragment>
            ))}
          </Fragment>
        )
      }

      return <Fragment key={globalIdx++}>{node}</Fragment>
    }

    const ret = (
      type:
        | string
        | React.FunctionComponent<{}>
        | React.ComponentClass<{}, any>,
      props: JSX.IntrinsicAttributes | any,
      children: React.ReactNode[] | React.ReactNode | undefined,
    ): JSX.Element => {
      if (Array.isArray(children)) {
        return React.createElement(
          type,
          props,
          <Fragment>
            {children.map((ch) => (
              <Fragment key={globalIdx++}>{retNode(ch)}</Fragment>
            ))}
          </Fragment>,
        )
      }
      if (children) {
        return React.createElement(
          type,
          {
            ...props,
          } as JSX.IntrinsicAttributes,
          children,
        )
      } else {
        return React.createElement(type, {
          ...props,
        } as JSX.IntrinsicAttributes)
      }
    }

    const allowedTypes = [
      'a',
      'b',
      'blockquote',
      'code',
      'del',
      'div',
      'em',
      'footer',
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'hr',
      'img',
      'input',
      'li',
      'ol',
      'p',
      'picture',
      'pre',
      'source',
      'span',
      'strong',
      'sub',
      'sup',
      'table',
      'tbody',
      'td',
      'th',
      'thead',
      'tr',
      'ul',
      'br',
      'figure',
      'figcaption',
    ]

    // clean up props, only pass down a limited set of safe props
    let trimProps: typeof props = {
      key: props?.key,
      style: props?.style,
      align: props?.align,
      className: props?.className,
      id: props?.id,
    }

    let trimChildren: typeof children | undefined = children

    const allowedCustomComponents = [
      // custom completely overridden components
      'embed',
      'iframe',
      'img',
      'pre',
      // our custom components
      'poll',
      'paywall',
      'subscribenow',
      ...(args.extraAllowedCustomComponents ?? []),
    ].map((s) => s.toLowerCase())

    // Custom components
    if (
      typeof type === 'function' &&
      type?.name &&
      allowedCustomComponents.includes(type?.name.toLowerCase())
    ) {
      const customComponentName = type?.name.toLowerCase()

      if (customComponentName === 'img') {
        if (typeof props.src !== 'string') {
          return <></>
        }
        if (
          !props.src.startsWith('https://') &&
          !props.src.startsWith('http://')
        ) {
          return <></>
        }

        trimProps.src = props?.src
        trimProps.height = props?.height
        trimProps.width = props?.width
        trimProps.alt = props?.alt
        trimChildren = undefined // can never have children
      }

      if (customComponentName === 'embed') {
        trimProps.src = props?.src
        trimChildren = undefined // can never have children
      }

      if (customComponentName === 'iframe') {
        trimProps.src = props?.src
        trimProps.title = props?.title
        trimProps.allow = props?.allow
        trimChildren = undefined // can never have children
      }

      if (['subscribenow', 'paywall'].includes(customComponentName)) {
        trimProps.article = args.article
        // Default to true in the client side renderer. When rendering posts for end-users, the premium content is already stripped out by the backend.
        trimProps.showPaywalledContent = args.showPaywalledContent ?? true
        trimProps.isSubscriber = args.isSubscriber ?? false
        trimProps.paidArticlesBenefitName = args.paidArticlesBenefitName
      }

      if (customComponentName === 'ad') {
        trimProps.subscriptionBenefitId = props?.subscriptionBenefitId
        trimProps.adsContext = args.adsContext
      }

      if (Array.isArray(trimChildren) && trimChildren.length === 0) {
        trimChildren = undefined
      }

      return ret(type, trimProps, trimChildren)
    }

    if (typeof type !== 'string') {
      return <></>
    }

    type = type.toLowerCase()

    if (!allowedTypes.includes(type)) {
      return <></>
    }

    if (args.defaultOverride) {
      return React.createElement(args.defaultOverride, {}, trimChildren)
    }

    if (type === 'a') {
      const href = props?.href
      if (
        typeof href === 'string' &&
        // Restricted protos.
        // Do not allow relative URLs.
        (href.startsWith('https://') ||
          href.startsWith('http://') ||
          href.startsWith('mailto://') ||
          href.startsWith('#'))
      ) {
        trimProps.href = props?.href
      } else {
        return <></>
      }
    }

    if (type === 'input') {
      if (props?.type !== 'checkbox') {
        return <></>
      }

      trimProps.type = 'checkbox'
      trimProps.checked = props?.checked
      trimProps.disabled = 'disabled'
      trimChildren = undefined // can never have children
    }

    if (type === 'source') {
      trimProps.media = props?.media
      trimProps.srcSet = props?.srcSet
    }

    if (Array.isArray(trimChildren) && trimChildren.length === 0) {
      trimChildren = undefined
    }

    return ret(type, trimProps, trimChildren)
  }

  return strictCreateElement
}

export const markdownOpts = {
  disableParsingRawHTML: false,
  forceBlock: true,
  overrides: {
    // Do not render by default.
    // The web and email renderers will register their corresponding implementations
    poll: () => <></>,
    Paywall: () => <></>,
    SubscribeNow: () => <></>,
    embed: () => <></>,
    iframe: () => <></>,

    // example style overrides
    img: (args: any) => <img {...args} style={{ maxWidth: '100%' }} />,
  },
} as const

// hasContent detects if node has any content recursively
// It walks through multiple layers of Fragments to see if there is any non-empty Fragment
export const hasContent = (node: React.ReactNode): boolean => {
  if (!node) {
    return false
  }

  if (node && typeof node === 'object' && 'key' in node) {
    if ('children' in node.props) {
      if (Array.isArray(node.props.children)) {
        for (const ch of node.props.children) {
          if (hasContent(ch)) {
            return true
          }
        }
      } else {
        const ch = node.props.children

        if (ReactIs.isFragment(ch)) {
          if (hasContent(ch)) {
            return true
          }
        } else {
          if (ReactIs.isElement(ch) || typeof ch === 'string') {
            return true
          }
        }
      }
    }
  }

  return false
}

export const firstChild = (
  node: React.ReactNode,
): React.ReactNode | undefined => {
  if (!node) {
    return undefined
  }

  if (node && Array.isArray(node)) {
    for (const ch of node) {
      const c = firstChild(ch)
      if (c !== undefined) {
        return c
      }
    }
    return undefined
  }

  if (ReactIs.isFragment(node)) {
    return firstChild(node.props.children)
  }

  if (node) {
    return node
  }

  return undefined
}
