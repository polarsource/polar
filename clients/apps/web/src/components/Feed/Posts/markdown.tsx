import { Article } from '@polar-sh/sdk'
import React from 'react'

export type RenderArticle = Pick<
  Article,
  'title' | 'body' | 'published_at' | 'byline' | 'organization'
>

// strictCreateElement removes unsupported types and attributes
export const wrapStrictCreateElement = (args: {
  article: RenderArticle
  showPaywalledContent?: boolean
  isSubscriber?: boolean
  defaultOverride?: React.FunctionComponent
}): ((
  type:
    | string
    | import('react').FunctionComponent<{}>
    | import('react').ComponentClass<{}, any>,
  props: JSX.IntrinsicAttributes | any,
  children: any,
) => JSX.Element) => {
  const strictCreateElement = (
    type:
      | string
      | import('react').FunctionComponent<{}>
      | import('react').ComponentClass<{}, any>,
    props: JSX.IntrinsicAttributes | any,
    children: any,
  ): JSX.Element => {
    const allowedTypes = [
      'a',
      'b',
      'blockquote',
      'code',
      'del',
      'div',
      'em',
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'img',
      'p',
      'pre',
      'strong',
      'sup',
      'table',
      'tbody',
      'th',
      'thead',
      'tr',
      'td',
      'li',
      'ul',
      'ol',
      'footer',
      'hr',
      'span',
      'input',
    ]

    // clean up props, only pass down a limited set of safe props
    let trimProps: typeof props = {
      key: props?.key,
      style: props?.style,
      align: props?.align,
      className: props?.className,
    }

    const allowedCustomComponents = [
      // custom completely overridden components
      'embed',
      'iframe',
      'img',
      'pre',
      // our custom components
      'poll',
      'paywall',
      'subscribenow',
    ]

    // Custom components
    if (
      typeof type === 'function' &&
      type?.name &&
      allowedCustomComponents.includes(type?.name.toLocaleLowerCase())
    ) {
      const customComponentName = type?.name.toLocaleLowerCase()

      if (customComponentName === 'img') {
        trimProps.src = props?.src
        trimProps.height = props?.height
        trimProps.width = props?.width
        children = undefined // can never have children
      }

      if (customComponentName === 'embed') {
        trimProps.src = props?.src
        children = undefined // can never have children
      }

      if (customComponentName === 'iframe') {
        trimProps.src = props?.src
        trimProps.title = props?.title
        trimProps.allow = props?.allow
        children = undefined // can never have children
      }

      return React.createElement(
        type,

        {
          ...trimProps,

          // Dependency inject article to all components
          article: args.article,

          // Default to true in the client side renderer. When rendering posts for end-users, the premium content is already stripped out by the backend.
          showPaywalledContent: args.showPaywalledContent ?? true,
          isSubscriber: args.isSubscriber ?? false,
        } as JSX.IntrinsicAttributes,
        children,
      )
    }

    if (typeof type !== 'string') {
      return <></>
    }

    type = type.toLowerCase()

    if (!allowedTypes.includes(type)) {
      return <></>
    }

    if (args.defaultOverride) {
      return React.createElement(args.defaultOverride, {}, props.children)
    }

    if (type === 'a') {
      const href = props?.href
      if (
        typeof href === 'string' &&
        // Restricted protos.
        // Do not allow relative URLs.
        (href.startsWith('https://') ||
          href.startsWith('http://') ||
          href.startsWith('mailto://'))
      ) {
        trimProps.href = props?.href
      } else {
        return <></>
      }
    }

    if (type === 'input') {
      if (props?.type !== 'checkbox') {
        return <></>
      }

      trimProps.type = 'checkbox'
      trimProps.checked = props?.checked
      trimProps.disabled = 'disabled'
      children = undefined // can never have children
    }

    return React.createElement(
      type,

      trimProps,
      children,
    )
  }

  return strictCreateElement
}

export const markdownOpts = {
  disableParsingRawHTML: false,
  overrides: {
    // Do not render by default.
    // The web and email renderers will register their corresponding implementations
    poll: () => <></>,
    Paywall: () => <></>,
    SubscribeNow: () => <></>,
    embed: () => <></>,
    iframe: () => <></>,

    // example style overrides
    img: (args: any) => <img {...args} style={{ maxWidth: '100%' }} />,
  },
} as const
