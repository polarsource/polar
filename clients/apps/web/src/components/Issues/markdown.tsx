import React from 'react'

export const generateMarkdownTitle = (
  title: string,
): React.ReactElement | React.ReactElement[] => {
  const matches: RegExpMatchArray[] = []
  for (const m of title.matchAll(/`([^`]*)`/g)) {
    matches.push(m)
  }

  if (matches.length === 0) {
    return <>{title}</>
  }

  let i = 0
  let offset = 0
  const nodes: React.ReactElement[] = []
  const matchCount = matches.length

  for (const match of matches) {
    if (match.index === undefined) {
      continue
    }

    i += 1
    if (offset < match.index) {
      nodes.push(
        <React.Fragment key={`0-${i}`}>
          {title.substring(offset, match.index)}
        </React.Fragment>,
      )
    }

    nodes.push(
      <span
        key={`1-${i}`}
        className="dark:bg-polar-700 rounded-md bg-gray-100 px-1.5 py-0.5"
      >
        {match[1]}
      </span>,
    )
    offset = match.index + match[0].length
    if (i === matchCount) {
      nodes.push(
        <React.Fragment key={`3-${i}`}>
          {title.substring(offset, title.length)}
        </React.Fragment>,
      )
    }
  }
  return nodes
}
