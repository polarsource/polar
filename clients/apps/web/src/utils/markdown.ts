export const firstImageUrlFromMarkdown = (markdown: string) => {
  return markdown.match(/!\[.*?\]\((.*?)\)/)?.[1]
}
