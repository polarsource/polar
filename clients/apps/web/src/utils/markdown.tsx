import { MarkdownToJSX } from 'markdown-to-jsx'

export const firstImageUrlFromMarkdown = (markdown: string): string | null => {
  const imagesMatch = markdown.match(/!\[.*?\]\((.*?)\)/)
  if (!imagesMatch) {
    return null
  }
  return imagesMatch[1]
}

export const markdownOptions: MarkdownToJSX.Options = {
  disableParsingRawHTML: true,
  forceBlock: false,
  overrides: {
    embed: () => <></>,
    iframe: () => <></>,
    a: (props) => (
      <a
        {...props}
        rel="noopener noreferrer nofollow"
        target="_blank"
        className="text-blue-400 hover:underline"
      />
    ),
  },
}
