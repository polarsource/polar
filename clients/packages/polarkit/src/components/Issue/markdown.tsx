import React from 'react'

export const generateMarkdownTitle = (
  title: string,
): React.ReactElement | React.ReactElement[] => {
  const matches: RegExpMatchArray[] = []
  for (const m of title.matchAll(/`([^`]*)`/g)) {
    matches.push(m)
  }

  if (matches.length === 0) {
    return <>{title}</>
  }

  let i = 0
  let offset = 0
  const nodes: React.ReactElement[] = []
  const matchCount = matches.length

  for (const match of matches) {
    if (match.index === undefined) {
      continue
    }

    i += 1
    if (offset < match.index) {
      nodes.push(
        <React.Fragment key={`0-${i}`}>
          {title.substring(offset, match.index)}
        </React.Fragment>,
      )
    }

    nodes.push(
      <span
        key={`1-${i}`}
        className="rounded-md bg-gray-100 py-0.5 px-1.5 dark:bg-gray-700"
      >
        {match[1]}
      </span>,
    )
    offset = match.index + match[0].length
    if (i === matchCount) {
      nodes.push(
        <React.Fragment key={`3-${i}`}>
          {title.substring(offset, title.length)}
        </React.Fragment>,
      )
    }
  }
  return nodes
}
