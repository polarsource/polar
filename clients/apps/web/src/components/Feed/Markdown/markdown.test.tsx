import { firstChild, hasContent } from './markdown'

describe('hasContent', () => {
  test('empty', () => {
    expect(hasContent(<></>)).toBe(false)
  })

  test('empty nested', () => {
    expect(
      hasContent(
        <>
          <></>
          <>
            <></>
          </>
        </>,
      ),
    ).toBe(false)
  })

  test('top', () => {
    expect(hasContent(<span>text</span>)).toBe(true)
  })

  test('content under fragment', () => {
    expect(
      hasContent(
        <>
          <span>text</span>
        </>,
      ),
    ).toBe(true)
  })
})

describe('firstChild', () => {
  test('empty', () => {
    expect(firstChild(<></>)).toMatchSnapshot()
  })

  test('empty nested', () => {
    expect(
      firstChild(
        <>
          <></>
          <>
            <></>
          </>
        </>,
      ),
    ).toMatchSnapshot()
  })

  test('top', () => {
    expect(firstChild(<span>text</span>)).toMatchSnapshot()
  })

  test('content under fragment', () => {
    expect(
      firstChild(
        <>
          <span>text</span>
        </>,
      ),
    ).toMatchSnapshot()
  })

  test('sibling content', () => {
    expect(
      firstChild(
        <>
          <span>t1</span>
          <span>t2</span>
          <span>t3</span>
        </>,
      ),
    ).toMatchSnapshot()
  })

  test('different nesting', () => {
    expect(
      firstChild(
        <>
          <>
            <span>t1</span>
            <span>t2</span>
          </>
          <span>t3</span>
        </>,
      ),
    ).toMatchSnapshot()
  })
})
