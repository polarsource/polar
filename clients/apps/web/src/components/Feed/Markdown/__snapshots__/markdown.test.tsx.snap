// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`firstChild content under fragment 1`] = `
<span>
  text
</span>
`;

exports[`firstChild different nesting 1`] = `
<span>
  t1
</span>
`;

exports[`firstChild empty 1`] = `undefined`;

exports[`firstChild empty nested 1`] = `undefined`;

exports[`firstChild sibling content 1`] = `
<span>
  t1
</span>
`;

exports[`firstChild top 1`] = `
<span>
  text
</span>
`;
